src/pipe.toit

```diff
@@ -0,0 +1,357 @@
+
+to command arg1 arg2 arg3 arg4:
+  return to [command, arg1, arg2, arg3, arg4]
+
+// Fork a program, and return its stdin pipe.  Uses PATH to find the program.
+// Can be passed either a command (with no arguments) as a string, or an array
+// of arguments, where the 0th argument is the command.
+to arguments:
+  if arguments is string:
+    return to [arguments]
+  pipe_ends := OpenPipe true
+  stdin := pipe_ends.fd
+  pipes := fork true stdin PIPE_INHERITED PIPE_INHERITED arguments[0] arguments
+  dont_wait_for pipes[3]
+  return pipe_ends
+
+from command arg1:
+  return from [command, arg1]
+
+from command arg1 arg2:
+  return from [command, arg1, arg2]
+
+from command arg1 arg2 arg3:
+  return from [command, arg1, arg2, arg3]
+
+from command arg1 arg2 arg3 arg4:
+  return from [command, arg1, arg2, arg3, arg4]
+
+// Fork a program, and return its stdout pipe.  Uses PATH to find the program.
+// Can be passed either a command (with no arguments) as a string, or an array
+// of arguments, where the 0th argument is the command.
+from arguments:
+  if arguments is string:
+    return from [arguments]
+  pipe_ends := OpenPipe false
+  stdout := pipe_ends.fd
+  pipes := fork true PIPE_INHERITED stdout PIPE_INHERITED arguments[0] arguments
+  dont_wait_for pipes[3]
+  return pipe_ends
+
+backticks command arg1:
+  return backticks [command, arg1]
+
+backticks command arg1 arg2:
+  return backticks [command, arg1, arg2]
+
+backticks command arg1 arg2 arg3:
+  return backticks [command, arg1, arg2, arg3]
+
+backticks command arg1 arg2 arg3 arg4:
+  return backticks [command, arg1, arg2, arg3, arg4]
+
+// Fork a program, and return the output from its stdout.  Uses PATH to find
+// the program.  Can be passed either a command (with no arguments) as a
+// string, or an array of arguments, where the 0th argument is the command.
+// Throws an exception if the program exits with a signal or a non-zero
+// exit value.
+backticks arguments:
+  if arguments is string:
+    return backticks [arguments]
+  pipe_ends := OpenPipe false
+  stdout := pipe_ends.fd
+  pipes := fork true PIPE_INHERITED stdout PIPE_INHERITED arguments[0] arguments
+  subprocess := pipes[3]
+  reader := reader.BufferedReader pipe_ends
+  reader.buffer_all
+  output := reader.read_string (reader.buffered)
+  exit_value := wait_for subprocess
+  pipe_ends.close
+  if (exit_value & PROCESS_SIGNALLED) != 0:
+    // Process crashed.
+    throw
+      "$arguments[0]: " +
+        signal_to_string (exit_value >> PROCESS_SIGNAL_SHIFT) & PROCESS_SIGNAL_MASK
+  exit_code := (exit_value >> PROCESS_EXIT_CODE_SHIFT) & PROCESS_EXIT_CODE_MASK
+  if exit_code != 0: throw "$arguments[0]: exit code $exit_code"
+  return output
+
+/**
+Returns the exit value of the process which can then be decoded into
+  exit code or signal number.
+
+See $exit_code and $exit_signal.
+*/
+wait_for subprocess:
+  wait_for_ subprocess
+  state := monitor.ResourceState_ process_resource_group_ subprocess
```

Comment:
Dispose the state after calling wait.

Reply:
https://github.com/toitlang/pkg-host/pull/2